<a href="https://colab.research.google.com/github/plancute/python/blob/master/deeplearning/code/07_Multi_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

5장 참 거짓 판단장치 : 로지스틱 회귀 (89-94p, part_2)

5-6. 여러 입력값을 갖는 로지스틱 회귀

플레이스홀더 : 입력값을 저장하는 일종의 그릇 <br>
tf.placeholder('데이터형','행렬의 차원','이름') 형태로 사용

In [ ]:
#-*- coding: utf-8 -*-
import numpy as np
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()

Instructions for updating:
non-resource variables are not supported in the long term


In [ ]:
# 실행할 때마다 같은 결과를 출력하기 위한 seed 값 설정
seed = 0
np.random.seed(seed)
tf.set_random_seed(seed)

In [ ]:
# x,y의 데이터 값
x_data = np.array([[2, 3],[4, 3],[6, 4],[8, 6],[10, 7],[12, 8],[14, 9]])
y_data = np.array([0, 0, 0, 1, 1, 1, 1]).reshape(7, 1)

In [ ]:
# 입력 값을 플래이스 홀더에 저장
X = tf.placeholder(tf.float64, shape=[None, 2])
Y = tf.placeholder(tf.float64, shape=[None, 1])

In [ ]:
# 기울기 a와 bias b의 값을 임의로 정함. 
# [2,1] 의미: 들어오는 값은 2개, 나가는 값은 1개
a = tf.Variable(tf.random_uniform([2,1], dtype=tf.float64)) 
b = tf.Variable(tf.random_uniform([1], dtype=tf.float64))

In [ ]:
# y 시그모이드 함수의 방정식을 세움 (matmul : 행렬곱 (a1x1 + a2x2) --> [a1,a2] * [x1,x2] )
y = tf.sigmoid(tf.matmul(X, a) + b)

In [ ]:
# 오차를 구하는 함수 (Y : 실제값, y : 예측값)
loss = -tf.reduce_mean(Y * tf.log(y) + (1 - Y) * tf.log(1 - y))

In [ ]:
# 학습률 
learning_rate = 0.1

In [ ]:
# loss 값을 최소로 하는 값 찾기
gradient_descent = tf.train.GradientDescentOptimizer(learning_rate).minimize(loss)

In [ ]:
predicted = tf.cast(y > 0.5, dtype=tf.float64)
accuracy = tf.reduce_mean(tf.cast(tf.equal(predicted, Y), dtype=tf.float64))


In [ ]:
# 학습
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    for i in range(3001):
        a_, b_, loss_, _ = sess.run([a, b, loss, gradient_descent], feed_dict={X: x_data, Y: y_data})
        if (i + 1) % 300 == 0:
            print("step=%d, a1=%.4f, a2=%.4f, b=%.4f, loss=%.4f" % (i + 1, a_[0], a_[1], b_, loss_))

    # 어떻게 활용하는가
    new_x = np.array([7, 6.]).reshape(1, 2)  #[7, 6]은 각각 공부 시간과 과외 수업수.
    new_y = sess.run(y, feed_dict={X: new_x})

    print("공부 시간: %d, 개인 과외 수: %d" % (new_x[:,0], new_x[:,1]))
    print("합격 가능성: %6.2f %%" % (new_y*100))



step=300, a1=0.7584, a2=-0.4500, b=-2.5621, loss=0.2587
step=600, a1=0.7741, a2=-0.2051, b=-4.0029, loss=0.1863
step=900, a1=0.6955, a2=0.1030, b=-5.0471, loss=0.1463
step=1200, a1=0.6003, a2=0.3952, b=-5.8751, loss=0.1202
step=1500, a1=0.5077, a2=0.6577, b=-6.5631, loss=0.1018
step=1800, a1=0.4230, a2=0.8905, b=-7.1524, loss=0.0881
step=2100, a1=0.3472, a2=1.0970, b=-7.6681, loss=0.0776
step=2400, a1=0.2797, a2=1.2809, b=-8.1266, loss=0.0693
step=2700, a1=0.2196, a2=1.4457, b=-8.5395, loss=0.0626
step=3000, a1=0.1660, a2=1.5943, b=-8.9151, loss=0.0571
공부 시간: 7, 개인 과외 수: 6
합격 가능성:  85.97 %


5-7. 실제값 적용하기 <br>
공부 시간: 7시간, 개인 과외 수: 6번 일때,
합격 가능성은?  85.97 %

5-8. 로지스틱 회귀에서 퍼셉트론으로 <br>
1957년 코넬 항공 연구소의 프랑크 로젠블라트 고안 및 발표 <br>
인공신경망, 오차역전파, 딥러닝의 기초